cabal-lint/src/Cabal/Lint/Summygroup.hs

```diff
@@ -0,0 +1,45 @@
+module Cabal.Lint.Summygroup (
```

Comment:
Whether something similar exists, I don't know. I wonder what the advantage of this class is to using modifiers like `Min a` or `Max a`? Also, set intersection is still assocative, so we won't lose `Semigroup` properties in the instsance for `Summygroup (Always a)`.

Reply:
I agree with @jorisdral, my gut reaction would be to use a `MonoidalMap` (defined locally or via `monoidal-containers`) and sth like
```haskell
newtype MonoidalMap k a = MonoidalMap (Map k a)

instance (Ord k, Semigroup a) => Semigroup (Map k a)
instance (Ord k, Semigroup a) => Monoid (Map k a) 
```
(could also be imported from `monoidal-containers`), and then define
```haskell
type MultiSet  a = MonoidalMap a (Sum Int)
type Always    a = MonoidalMap a (Min Int)
type Sometimes a = MonoidalMap a (Max Int)
```
which are pairwise `coerce`ible.

As for the use case of writing generic code that is polymorphic in `(Semigroup m, Summygroup m)`: This would probably first require finding out sensible laws how `(<>)` and `(<+>)` shall interact (i.e. some distributivity, otherwise, it seems too ad-hoc). Maybe they form a [`Semiring`](https://hackage.haskell.org/package/semirings-0.6/docs/Data-Semiring.html#t:Semiring)?

In any case, parameterizing `goTree`/`goBranch` by an extra argument `m -> m -> m` instead of `Summygroup` (that is instantiated with `coerce (<>) @...`) is another alternative.

Reply:
I.e. maybe we can lift [tropical semirings](https://en.wikipedia.org/wiki/Tropical_semiring) to `Map`s? cc @bartfrenk

Reply:
Esgen and I chatted about this on our call today. See the new Haddock on the class for what we came up with.

Reply:
> Whether something similar exists, I don't know. I wonder what the advantage of this class is to using modifiers like `Min a` or `Max a`.

1) My `a` types are not numbers, so I can't rely on `Min` or `Max` per se. My types are sets, maps of non-empty sets, tuples thereof, etc.

2) So maybe I define my own more general newtypes `Minny` and `Maxxy`. But at that point, I think I'd be relying on `(Semigroup . Minny)` or whatever in just the same way as I'm relying on `Summygroup`, wouldn't I?

The `a` is the result of the tool's various analyses of the Cabal AST. The way I think about it as that that analysis codomain needs combinators for both juxtaposition (I use `Monoid` for this) and combination of branches (I use `Summygroup` for this).